packages/babel-parser/src/plugins/typescript/index.ts

```diff
@@ -1092,47 +1091,87 @@ export default (superClass: ClassWithMixin<typeof Parser, IJSXParserMixin>) =>
       const { startLoc } = this.state;
 
       const rest = this.eat(tt.ellipsis);
-      const isKeyword = tokenIsKeyword(this.state.type);
-      let type: N.TsNamedTupleMember | N.TsType | N.Identifier = isKeyword
-        ? this.parseIdentifier(true)
-        : this.tsParseType();
-      const optional = this.eat(tt.question);
-      const labeled = this.eat(tt.colon);
 
-      if (labeled) {
-        const labeledNode = this.startNodeAtNode<N.TsNamedTupleMember>(type);
-        labeledNode.optional = optional;
+      let labeled: boolean;
+      let label: N.Identifier;
+      let optional: boolean;
+      let type: N.TsNamedTupleMember | N.TsType;
+
+      const isWord = tokenIsKeywordOrIdentifier(this.state.type);
+      if (isWord && this.lookaheadCharCode() === charCodes.colon) {
+        labeled = true;
+        optional = false;
+        label = this.parseIdentifier(true);
+        this.expect(tt.colon);
+        type = this.tsParseType();
+      } else if (
+        isWord &&
+        this.lookaheadCharCode() === charCodes.questionMark
+      ) {
+        optional = true;
+        const startLoc = this.state.startLoc;
+        const wordName = this.state.value;
+        const typeOrLabel = this.tsParseNonArrayType();
+
+        if (this.lookaheadCharCode() === charCodes.colon) {
+          labeled = true;
+          label = this.createIdentifier(
+            this.startNodeAt<N.Identifier>(startLoc),
+            wordName,
+          );
+          this.expect(tt.question);
+          this.expect(tt.colon);
+          type = this.tsParseType();
+        } else {
+          labeled = false;
+          type = typeOrLabel;
+          this.expect(tt.question);
+        }
+      } else {
+        type = this.tsParseType();
+        optional = this.eat(tt.question);
+        labeled = this.eat(tt.colon);
```

Comment:
When can this be followed by a colon?

Reply:
In case we have invalid label in tuple followed by colon, which fails `tokenIsKeywordOrIdentifier` check?

I'm sorry to bother you, this issue seemed easier than it really is for me

Reply:
I was thinking about the case described in "tuple-invalid-label-1" test: `type T = [x.y: A];` which should throw error `TSErrors.InvalidTupleMemberLabel` to be passed

Reply:
Ok :+1: Maybe add a comment in the code with that example, explaining that it can only happen in that case.
Also, probably we can move the `TSErrors.InvalidTupleMemberLabel` inside this block so that it's throw in the same place as where we parse the invalid code.